### ISSUE TITLE
TS Server command to get inverse dependencies of file

### ISSUE THREAD

**Issue by mjbvz** [microsoft/TypeScript#36664]
For https://github.com/microsoft/vscode/issues/13953

## Background
Project wide diagnostics are the most requested VS Code JS/TS feature. However, we've been blocked on implementing them by the fact that TS server is single threaded. This means that computing these diagnostics could interrupt normal user operations like completions.

We currently recommend that people use tasks to get around this limitation (since tasks always run as separate process) however few people actually use tasks.

With https://github.com/microsoft/vscode/commit/f0942786b4ee338f02449e2e5751091e028eb898, I've added a new setting that spins up a new TS Server just for diagnostics. This accomplishes much the same thing as using a task.

## Problem
In the prototype, I'm using the `geterrForProject` command. This works ok for smaller projects but triggers a huge number of events for larger projects like the VS Code source. Instead, we need an API that: given a file, returns a list of inverse dependencies (i.e. the files that need to be rechecked after the file changes)

I believe that this API more or less already exists with `compileOnSaveAffectedFileList`? But that command only works if `compileOnSave` is enabled. We'd like to enable project wide error reporting even when users have not enabled compileOnSave.

/cc @sheetalkamat For background on `compileOnSaveAffectedFileList`
/cc @amcasey As you suggested using a separate diagnostics only server

**Comment by mjbvz** [comment]
Related question: before calling `geterr` on a file do I always have to use `updateOpen` on it? Right now it seems like I have to explicitly open files against TS server in order to get back errors for them.

**Comment by RyanCavanaugh** [comment]
@sheetalkamat thoughts?

**Comment by uniqueiniquity** [comment]
FWIW - @mjbvz met about this and I'm working on prototyping project-wide diagnostics more broadly than this particular ask

**Comment by DanielRosenwasser** [comment]
#15168 might be a relevant discussion point for this issue too.

**Comment by DanielRosenwasser** [comment]
> Instead, we need an API that: given a file, returns a list of inverse dependencies (i.e. the files that need to be rechecked after the file changes)

Yes, the first thing you want to see is errors from the current file ("what's affected in this file?") and then that file's dependents ("what did this file affect?"); however, when would you query for errors for the dependencies of the open files? (i.e. "what's wrong with whatever this file depends on?")

**Comment by sschafercdd** [comment]
I just turned on `tsserver.experimental.enableProjectDiagnostics` and it seems to give me what I expect. Files with Typescript errors show as red in the Explorer. So that's nice.

But it does seem slow, and in general the error checking is just a bit sluggish. I might fix a syntax error but have to wait a second or two before it registers.

It does show the `node_modules` folder in red and appears to be checking that for errors despite that folder being excluded in `tsconfig.json`.

Seems like this issue has been open for a while, and it also seems like it would be fairly trivial to fix. So...can we get a fix for this?

**Comment by rodrigocipriani** [comment]
> tsserver.experimental.enableProjectDiagnostics

Small fix, the right string on the settings.json is `"typescript.tsserver.experimental.enableProjectDiagnostics": true`

**Comment by dallman2** [comment]
Is there an expected time to completion on this issue? It would be nice to see what breaks in my codebase when changing lint rules, or just refactoring types...

**Comment by jbouduin** [comment]
First question: any progress on this?

Second question: did I see it correctly, that when setting enableProjectDiagnostics to true, its starts diagnostics on node_modules also ? If this is the case, there should be a way to exclude/include directories. (Maybe there is, and I am just not aware of it)

**Comment by jerrygreen** [comment]
> did I see it correctly, that when setting enableProjectDiagnostics to true, its starts diagnostics on node_modules also

Yes and no. Normally it shouldn't, but sometimes it is. The setting is experimental for years, it works unreliably. You can try the setting and hope for the best.

I tried it, and personally I loved it... At first. But sometimes I had to restart VSCode so it stops looking into node_modules. Then I tired of doing it regularly, so I ended up ditching the setting. It just doesn't work reliably enough. I hope they debug the setting someday, because in the theory it's a cool setting, but it has bugs, and for all this time, they weren't fixed. Sadge.

**Comment by RyanCavanaugh** [comment]
> "typescript.tsserver.experimental.enableProjectDiagnostics": true

This is the correct setting to enable. I will ask VS Code if we can move this out of "experimental" since it's been there a *long* time.

> but it has bugs, and for all this time, they weren't fixed. Sadge.

Which ones?

**Comment by louwers** [comment]
https://github.com/microsoft/vscode/issues/117732 was closed as a duplicate of this issue, but the issue still seems to be present.

**Comment by FunctionDJ** [comment]
@ ryan
> This is the correct setting to enable. I will ask VS Code if we can move this out of "experimental" since it's been there a _long_ time.

please do! this would progress several issues in other projects that won't support `enableProjectDiagnostics` because it's experimental, like denoland/vscode_deno/issues/66

**Comment by grundmanise** [comment]
Just wanted to leave a comment here since https://github.com/microsoft/vscode/issues/117732 is locked. Even today, I still have an issue where it reports node module problems even when they're excluded from files. But going into each file with a problem makes them disappear – until you restart.
